### PR TITLE
Add fusion pass

### PR DESCRIPTION
**Pull request by jbruestle**

### CODE REVIEW

File: pmlc/dialect/pxa/analysis/strides.cc [hunk 1]
```diff
@@ -1,11 +1,11 @@
 // Copyright 2020 Intel Corporation
 
-#include "pmlc/dialect/pxa/analysis/strides.h"
-
 #include <map>
 #include <string>
 #include <vector>
 
+#include "pmlc/dialect/pxa/analysis/strides.h"
```
Comment (flaub) [inline pmlc/dialect/pxa/analysis/strides.cc:7]: This include should go first.

File: pmlc/dialect/pxa/analysis/strides.cc [hunk 2]
```diff
@@ -115,10 +115,13 @@ Optional<StrideInfo> computeStrideInfo(Value expr) {
   }
 
   // Try for the affine apply case
-  if (auto op = dyn_cast<AffineApplyOp>(expr.getDefiningOp()))
+  if (auto op = dyn_cast_or_null<AffineApplyOp>(expr.getDefiningOp()))
     return computeStrideInfo(op.getAffineMap().getResult(0),
                              op.getMapOperands());
 
+  IVLOG(1, "Failed stride info: op = "
+               << expr.getDefiningOp()->getName().getStringRef().str());
```
Comment (flaub) [inline pmlc/dialect/pxa/analysis/strides.cc:123]: Could use `mlir::debugString(op)` here

File: pmlc/dialect/pxa/transforms/BUILD [hunk 3]
```diff
@@ -23,6 +23,7 @@ plaidml_cc_library(
     name = "transforms",
     srcs = [
         "autotile.cc",
+	"fusion.cc",
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/BUILD:26]: Whitespace

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+
+namespace {
+
+using WriteRead = std::pair<AffineReduceOp, AffineLoadOp>;
+using WriteWrite = std::pair<AffineReduceOp, AffineReduceOp>;
+
+struct FusionInfo {
+  // The parallel for ops
+  AffineParallelOp apA;
+  AffineParallelOp apB;
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:27]: Naming: could we use `consumer` and `producer`?

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+  // The loop sizes for apA and apB
+  llvm::SmallVector<int64_t, 4> sizesA;
+  llvm::SmallVector<int64_t, 4> sizesB;
+  // The load and store ops
+  llvm::SmallVector<WriteRead, 4> writeReads;
+  llvm::SmallVector<WriteWrite, 4> writeWrites;
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:33]: Naming: use `writeAfterReads` and `writeAfterWrites` or `wars` and `waws`

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+          mlir::cast<mlir::AffineYieldOp>(op.getBody()->getTerminator());
+      return findSourceWrite(retOp.getOperand(opRes.getResultNumber()));
+    }
+    return AffineReduceOp();
+  }
+
+  // Helper method to remove elements from a stride info that are not part of
+  // a specific block's arguments.
+  static void cleanStrideInfo(Block *block, StrideInfo &si) {
+    for (auto it = si.strides.begin(); it != si.strides.end();) {
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:64]: Can we use `llvm::make_early_inc_range` here?

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+      // If both are empty, nothing to do
+      if (sa.strides.size() == 0 && sb.strides.size() == 0)
+        continue;
+      // TODO: Thses should be fixed by canonicalization
+      if (sa.strides.size() == 0 && sb.strides.size() == 1 &&
+          sizesB[sb.strides.begin()->first.getArgNumber()] == 1)
+        continue;
+      if (sb.strides.size() == 0 && sa.strides.size() == 1 &&
+          sizesA[sa.strides.begin()->first.getArgNumber()] == 1)
+        continue;
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:122]: Do we still need this post-canonicalization?

File: pmlc/dialect/pxa/transforms/fusion.cc (hunk 4 above)
Comment (jbruestle) [reply pmlc/dialect/pxa/transforms/fusion.cc:27]: Producer/consumer sounds too much like it's limited to a single readAfterWrite to me.  Additionally, for most things, we handle the two block similarly, and there are a number of places where I needs someVarA and someVarB, so I'd like to keep the names shortish and regular looking.

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+
+    // Extract the per-dimension strides for the two ops
+    llvm::SmallVector<StrideInfo, 4> stridesA;
+    llvm::SmallVector<StrideInfo, 4> stridesB;
+    if (!getStrides(stridesA, opA, apA))
+      return false;
+    if (!getStrides(stridesB, opB, apB))
+      return false;
+
+    assert(stridesA.size() == stridesB.size());
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:101]: Add "&&" message

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+
+namespace {
+
+using WriteRead = std::pair<AffineReduceOp, AffineLoadOp>;
+using WriteWrite = std::pair<AffineReduceOp, AffineReduceOp>;
+
+struct FusionInfo {
+  // The parallel for ops
+  AffineParallelOp apA;
+  AffineParallelOp apB;
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:27]: It might be a little easier to read by expanding `apA` to `affineParallelA`

Comment (flaub) [reply pmlc/dialect/pxa/transforms/fusion.cc:27]: Actually, could we place all the state for a particular `affine.parallel` in its own struct?

like:
```
struct FusionAffineParallelInfo {
  AffineParallelOp op;
  llvm::SmallVector<int64_t, 4> sizes;
  llvm::SmallVector<int64_t, 8> tile;
};
```

File: pmlc/dialect/pxa/transforms/fusion.cc [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+                   AffineMapAttr::get(newUpper));
+      apOp.setAttr(AffineParallelOp::getStepsAttrName(),
+                   builder.getI64ArrayAttr(newSteps));
+    };
+    fixupLoops(apA, aToNew);
+    fixupLoops(apB, bToNew);
+    // PRINT!
+    // mlir::OpPrintingFlags flags;
+    // flags.printGenericOpForm();
+    // apC.getOperation()->print(llvm::errs(), flags);
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/fusion.cc:346]: Not sure if we should keep this around

File: pmlc/dialect/pxa/transforms/fusion.cc (hunk 9 above)
Comment (jbruestle) [reply pmlc/dialect/pxa/transforms/fusion.cc:27]: Probably a good idea, I'll try to refactor to something like that when I add the actual tiling support, think I will leave it for today though if that's OK.

**Pull request merged**